src/main.cc

```diff
@@ -135,12 +135,81 @@ sig_checkpoint(ev_loop * /* loop */, struct ev_signal * /* w */,
 	fiber_wakeup(f);
 }
 
+static void
+trigger_fiber_run_on_timeout(ev_loop *loop, ev_timer *watcher, int revents)
+{
+	(void)loop;
+	(void)revents;
+	bool *expired = (bool *)watcher->data;
+	assert(expired != NULL);
+	*expired = true;
+}
+
+static int
+trigger_fiber_f(va_list ap)
```

Comment:
All these helper functions should be prefixed `on_shutdown_` so that it's clear what they are related to.

Anyway, this isn't what I meant by "inlining `trigger_fiber_run` in `on_shutdown_f`". I meant that we need to rework `on_shutdown_f` so that it runs user callbacks in fibers using only public trigger API (`trigger_run`), without accessing trigger internals. I think we can do this by adding a wrapper function around a user callback that would call the user callback in a fiber. We would put the wrapper functions in the trigger list instead of user callbacks. Can we teach `lbox_trigger_reset` to do this? If you can't do this, better leave `fiber_trigger_run` as is.

Also, this had better be done before the main patch, which fixes `trigger_run` issues.

Reply:
>  I meant that we need to rework on_shutdown_f so that it runs user callbacks in fibers using only public trigger API (trigger_run), without accessing trigger internals.

Oh, now I see. Misunderstood that at first.
Anyway, I tried your approach and it doesn't work out too well.
As discussed verbally, I left `trigger_fiber_run` where it was and fixed the issue there.
Also found and fixed typos in `trigger_run` and `trigger_run_reverse`.

<details><summary>diff</summary>

```diff
diff --git a/src/lib/core/trigger.cc b/src/lib/core/trigger.cc
index deb838db7..23bc26ed6 100644
--- a/src/lib/core/trigger.cc
+++ b/src/lib/core/trigger.cc
@@ -145,7 +145,7 @@ trigger_run(struct rlist *list, void *event)
 	struct trigger *trigger;
 	rlist_foreach_entry(trigger, list, link) {
 		if (run_list_put_trigger(&run_list, trigger) != 0) {
-			run_list_clear(list);
+			run_list_clear(&run_list);
 			return -1;
 		}
 	}
@@ -159,7 +159,7 @@ trigger_run_reverse(struct rlist *list, void *event)
 	struct trigger *trigger;
 	rlist_foreach_entry_reverse(trigger, list, link) {
 		if (run_list_put_trigger(&run_list, trigger) != 0) {
-			run_list_clear(list);
+			run_list_clear(&run_list);
 			return -1;
 		}
 	}
@@ -181,7 +181,7 @@ trigger_clear(struct trigger *trigger)
 int
 trigger_fiber_run(struct rlist *list, void *event, double timeout)
 {
-	struct trigger *trigger, *tmp;
+	struct trigger *trigger;
 	unsigned trigger_count = 0;
 	struct region *region = &fiber()->gc;
 	RegionGuard guard(region);
@@ -198,6 +198,14 @@ trigger_fiber_run(struct rlist *list, void *event, double timeout)
 		return -1;
 	}
 
+	RLIST_HEAD(run_list);
+	rlist_foreach_entry(trigger, list, link) {
+		if (run_list_put_trigger(&run_list, trigger) != 0) {
+			run_list_clear(&run_list);
+			return -1;
+		}
+	}
+
 	bool expired = false;
 	struct ev_timer timer;
 	ev_timer_init(&timer, trigger_fiber_run_timeout, timeout, 0);
@@ -211,7 +219,8 @@ trigger_fiber_run(struct rlist *list, void *event, double timeout)
 	ev_timer_start(loop(), &timer);
 
 	unsigned current_fiber = 0;
-	rlist_foreach_entry_safe(trigger, list, link, tmp) {
+	while (!rlist_empty(&run_list)) {
+		trigger = run_list_take_trigger(&run_list);
 		char name[FIBER_NAME_INLINE];
 		snprintf(name, FIBER_NAME_INLINE,
 			 "trigger_fiber%d", current_fiber);
@@ -222,6 +231,7 @@ trigger_fiber_run(struct rlist *list, void *event, double timeout)
 			current_fiber++;
 		} else {
 			ev_timer_stop(loop(), &timer);
+			run_list_clear(&run_list);
 			return -1;
 		}
 	}

```

</details>